File: leetcode_0211/java/leetcode_0211.java
public class leetcode_0211 {
    class WordDictionary {
        class Node{
            char val;
            //判断是否是一个完整单词
            boolean end;
            Node[] nexts;
            Node(char val){this.val = val;end = false;nexts = new Node[26];}
        }
        //由于需要提前知道前一个元素，才能判断下一个元素的值 这里创建一个虚拟头节点
        private Node dummyHead;
        public WordDictionary() {
            dummyHead = new Node(' ');
        }

        public void addWord(String word) {
            Node cur = dummyHead;
            for(int i = 0;i < word.length();i++){
                char ch = word.charAt(i);
                int num = ch - 'a';
                if(cur.nexts[num] == null){
                    cur.nexts[num] = new Node(ch);
                    cur = cur.nexts[num];
                }else{
                    cur = cur.nexts[num];
                }
                if(i == word.length() - 1)
                    cur.end = true;
            }
        }

        public boolean search(String word) {
            return searchHelper(dummyHead,word);
        }
        private boolean searchHelper(Node cur,String word){
            if(word.length() == 0)
                return true;
            // if(cur == null && word.length() == 0)
            //     return false;
            for(int i = 0;i < word.length();i++){
                int num = word.charAt(i) - 'a';
                if(num >= 0 && num < 26 && cur.nexts[num] == null)
                    return false;
                else if(num >= 0 && num < 26 && cur.nexts[num] != null)
                    cur = cur.nexts[num];
                else{
                    //说明字符是'.'
                    //进行dfs
                    for(int j = 0;j < 26;j++){
                        if(i + 1 < word.length()){
                            if(cur.nexts[j] != null && searchHelper(cur.nexts[j],word.substring(i+1,word.length())))
                                return true;
                        }else{
                            if(cur.nexts[j] != null && cur.nexts[j].end)
                                return true;
                        }
                    }
                    return false;
                }
            }
            return (cur == null)?true:cur.end;
        }
    }

    /**
     * Your WordDictionary object will be instantiated and called as such:
     * WordDictionary obj = new WordDictionary();
     * obj.addWord(word);
     * boolean param_2 = obj.search(word);
     */
    public static void main(String[] args) {
        leetcode_0211.WordDictionary wd = new leetcode_0211().new WordDictionary();
        wd.addWord("at");
        wd.addWord("and");
        wd.addWord("an");
        wd.addWord("add");
        System.out.println(wd.search("a"));
        //result
        //false  不是单词终点
        System.out.println(wd.search(".at"));
        //result
        //false
        wd.addWord("bat");
        //result
        //true
        System.out.println(wd.search(".at"));
        //result
        //true
        System.out.println(wd.search("an."));
        //result
        //true
        System.out.println(wd.search("a.d."));
        //result
        //false
    }
}

File: leetcode_0211/cpp/leetcode_0211.cpp
//
// Created by wzy on 2022/4/22.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct Node{
    char val;
    bool end;
    vector<Node*> nexts;
    Node(char val):val(val),end(false),nexts(vector<Node*>(26,nullptr)){};
};
class WordDictionary {
public:
    WordDictionary() {
        //思路：Trie 前缀树 对于.的判断使用递归处理
        dummyHead = new Node(' ');
    }
    ~WordDictionary() = default;

    void addWord(string word) {
        Node* cur = dummyHead;
        for(int i = 0;i < word.size();i++){
            int num = word[i] - 'a';
            if(!cur->nexts[num]){
                cur->nexts[num] = new Node(word[i]);
                cur = cur->nexts[num];
            }else{
                cur = cur->nexts[num];
            }
            if(i == word.size()-1)
                cur->end = true;
        }
    }

    bool search(string word) {
        return searchHelper(dummyHead,word);
    }
private:
    Node* dummyHead;
    bool searchHelper(Node* cur,string word){
        if(!word.size() )
            return true;
        if(!cur && word.size())
            return false;
        for(int i = 0;i < word.size();i++){
            int num = word[i] - 'a';
            if(num >= 0 && num < 26 && !cur->nexts[num])
                return false;
            else if(num >= 0 && num < 26 && cur->nexts[num]){
                cur = cur->nexts[num];
            }else{
                for(int j = 0;j < 26;j++){
                    if(i + 1 < word.size()){
                        if(cur->nexts[j] && searchHelper(cur->nexts[j],word.substr(i+1,word.size()-i-1)))
                            return true;
                    }
                    else{
                        //i+1 == word.size()
                        if(cur->nexts[j] && cur->nexts[j]->end)
                            return true;
                    }
                }
                return false;
            }
        }
        //如果都是字母从这里结束，判断节点是否有完整单词标志
        return (!cur)?true:cur->end;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

int main(){
    WordDictionary* wd = new WordDictionary();
    wd->addWord("at");
    wd->addWord("and");
    wd->addWord("an");
    wd->addWord("add");
    cout<<boolalpha<<wd->search("a")<<endl;
    //result
    //false  不是单词终点
    cout<<boolalpha<<wd->search(".at")<<endl;
    //result
    //false
    wd->addWord("bat");
    cout<<boolalpha<<wd->search(".at")<<endl;
    //result
    //true
    cout<<boolalpha<<wd->search("an.")<<endl;
    //result
    //true
    cout<<boolalpha<<wd->search("a.d.")<<endl;
    //result
    //false

}